### DISCUSSION TITLE
Code Style Checks as GitHub Annotations (for PRs from a fork)

### DISCUSSION THREAD

**Discussion by tarasio-mirror** [pytorch/ignite#1698] [Ideas]
This idea is a kind of result of my recent research on GH Actions.

Due to GitHub Actions restrictions on forked repositories, our `code-style` workflow does not work for PRs raised from forks. 

Possible solutions are:

## Accept it

Conditionally ignore failing step. Ask contributors to turn on GitHub Actions for their forks.

## Adapt it

Another solution that I want to propose and discuss is to use formatters in check mode, and use those checks for pointing out code style issues using GH Action Logs and Annotations in PR events from a fork. All the auto-formatting still will be available for non-fork events.

For example, I've created the MVP version of `pycodestyle` for GitHub Actions workflows. (we cannot use `autopep8` in check mode, so we need to replace it with `pycodestyle`) 

Here is what I am trying to achieve (results from a dummy demo repo):

After style check we have list of errors and warnings:
![Screenshot from 2021-02-24 22-00-20](https://user-images.githubusercontent.com/30748114/109059091-7f000800-76ec-11eb-9776-36babf9f94de.png)

From that you can jump to a specific file or line:
![Screenshot from 2021-02-24 22-00-06](https://user-images.githubusercontent.com/30748114/109059170-9939e600-76ec-11eb-9358-28144f1d9787.png)

Logs are grouped for each file
![Screenshot from 2021-02-24 22-01-18](https://user-images.githubusercontent.com/30748114/109059301-c5edfd80-76ec-11eb-9dcf-551bd8491e19.png)

After unfolding you will find detailed info on error
![Screenshot from 2021-02-24 22-01-27](https://user-images.githubusercontent.com/30748114/109059330-d00ffc00-76ec-11eb-935b-6e243032ce8b.png)

And here is demo run on ignite repo and dummy file "fixme.py":
https://github.com/pytorch/ignite/actions/runs/597334732

So, currently it works only for `pycodestyle`, but I think we can achieve similar behavior for `black` and `isort`

**Comment by jeffydc** [comment]
Regarding about code style checking, can we have 2 bash scripts one for formatting, one for linting ?
We can reuse in the CI too.

So that if some don't use pre-commit, they can run those scripts before commit without copying and pasting commands from `unit-tests.yml` every time.
